### PR TITLE
Vim Mode 4 - Core functionality for Vim Mode

### PR DESCRIPTION
**Pull request by tntmarket**
This is part of a larger change to support vim style navigation (#63)

This actually introduces the vim commands

* Basic navigation
* Insert mode
* Visual mode

It also has abstractions on top of Roam that make it more convenient to implement Vim Mode. The main abstractions that were introduced are:

* Panel - a set of blocks that you navigate between. Analogous to a "Vim Window"
* Roam Events - listeners to roam user actions such as toggling the right panel

I didn't integrate this with `core/roam` yet, since I wanted to get feedback on whether they were the right abstractions first.

### CODE REVIEW

**Comment by Stvad** [comment]
haven't looked at it in detail yet, but:
* great stuff!
* most of the function probably shouldn't be inline with the shortcut (so they can potentially be reused elsewhere)

**Comment by tntmarket** [comment]
Are you thinking something more like this?

```javascript
nmap('i', 'Click Selection', clickSelection),
nmap('a', 'Click Selection and Go-to End of Line', clickSelectionAndGotoEndOfLine),
nmap('O', 'Insert Block Before', insertBlockBefore),
nmap('o', 'Insert Block After', insertBlockBefore),
```

**Comment by Stvad** [comment]
yeah

**Comment by tntmarket** [comment]
Cool, would you be ok with extracting them as top-level functions in the files themselves for now? Then people can export and move them to a different file if they want to reuse them.

**Comment by Stvad** [comment]
yeah, that's probably fine. though if you see opportunity to move a function to a better fitting place - take it :)

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import {BlockElement} from 'src/core/features/vim-mode/roam/roam-block'
+import {onSelectorChange, waitForSelectorToExist} from 'src/core/common/mutation-observer'
+import {Selectors} from 'src/core/roam/selectors'
+import {assumeExists} from 'src/core/common/assert'
+
+type DisconnectFn = () => void
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts]: I think this one was already defined elsewhere?

File: src/ts/core/features/vim-mode/roam/roam-block.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * The generically reusable parts of this should probably move to core/roam
+ */
+export class RoamBlock {
+    private blockId: BlockId
+
+    constructor(blockId: BlockId) {
+        this.blockId = blockId
+    }
+
+    element(): BlockElement {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-block.ts]: should it be a property?

**Comment by Stvad** [comment]
Another general comment here - I feel like for commands where we start in `normal` mode we should (most of the time) end up in the normal mode at the end again (even if executing command requires us to go to insert mode temporarily)

File: src/ts/core/features/vim-mode/roam/roam-block.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        return assumeExists(document.getElementById(this.blockId))
+    }
+
+    async focus() {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-block.ts]: `focus` evokes for me something like "select in normal mode" this feels more like `edit` or `activate`

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+    document.addEventListener(eventType, handleBlockEvent)
+    return () => document.removeEventListener(eventType, handleBlockEvent)
+}
+
+/**
+ * Various helpers for detecting user actions
+ *
+ * The generically reusable parts of this should probably move to core/roam
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts:20]: ❤️

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ */
+export const RoamEvent = {
+    // Triggers when the right panel is shown or hidden
+    onRightPanelToggle(handler: (isRightPanelOn: boolean) => void): DisconnectFn {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts]: wonder if `sidebar` is a better wording as that's what conventionally used? or is panel wording better within the context of having central panel/etc?

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                handler(true)
+            } else {
+                handler(false)
+            }
+        })
+    },
+
+    // Triggers when opening or closing an article in the right panel
+    onRightPanelChange(handler: () => void): DisconnectFn {
+        let stopObserving = () => {}
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts:33]: nit: use defined type

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        return RoamEvent.onRightPanelToggle(isRightPanelOn => {
+            if (isRightPanelOn) {
+                stopObserving = onSelectorChange(Selectors.sidebarContent, handler)
+            } else {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts:37]: do we even care if we keep observing while it's hidden?

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                stopObserving()
+            }
+        })
+    },
+
+    onFocusBlock(handler: (element: BlockElement) => void): DisconnectFn {
+        return onBlockEvent('focusin', handler)
+    },
+
+    onBlurBlock(handler: () => void): DisconnectFn {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts]: should we pass the block to the handler?

File: src/ts/core/features/vim-mode/roam/roam-event.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        return RoamEvent.onRightPanelToggle(isRightPanelOn => {
+            if (isRightPanelOn) {
+                stopObserving = onSelectorChange(Selectors.sidebarContent, handler)
+            } else {
+                stopObserving()
+            }
+        })
+    },
+
+    onFocusBlock(handler: (element: BlockElement) => void): DisconnectFn {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-event.ts]: when we focus the element does it get transformed into the textbox? or we can focus without clicking (i.e. normal mode) and so this is orthogonal to going to edit mode?

File: src/ts/core/features/vim-mode/roam/roam-highlight.ts [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * A "Highlight" is the native roam selection you get when
+ * selecting of blocks.
+ *
+ * Native highlights are reused in order to simulate Vim's "visual mode".
+ *
+ * The generically reusable parts of this should probably move to core/roam
+ */
+export class RoamHighlight {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-highlight.ts]: given that it has only static functions should this be just an object?

Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-highlight.ts]: also probably want to expose all highlighted blocks function as well

File: src/ts/core/features/vim-mode/roam/roam-highlight.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    static firstBlockElement(): BlockElement {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-highlight.ts]: probably just `first` and `last`

**Review by Stvad — COMMENTED** [review]
pushing a partial review. gonna take a break now.  
as a general comment for the cases when you think a particular functionality should go to `core/roam` - please move it there ;)

**Comment by Stvad** [comment]
btw I believe `g g` and `G` mappings got lost in translation somewhere :)

**Comment by tntmarket** [comment]
I'll work on putting them back.

Initially, I cut them out because the implementation where I was simulating up/down 100 times seemed dubious :p

File: src/ts/core/features/vim-mode/roam/roam-block.ts (hunk 3 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-block.ts]: I'll use `edit`

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 5 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-event.ts]: Actually I'll use `sidebar`, since `RoamEvent` could be useful regardless of whether we're using the "panel" abstraction

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 7 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-event.ts:37]: I currently observe when divs appear/disappear within `Selectors.sidebarContent`. `Selectors.sidebarContent` doesn't exist when the sidebar is closed however.

I could observe for appearances of `Selectors.sidebarContent > div` within `body`, but that would require me to use a `subtree: true` mutation observer, which I have been trying to avoid (for performance and explicitness)

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 9 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-event.ts]: So my current vocabulary is:

1. "Focus" (to be renamed "Edit" as suggested) -> Clicking into a block in order to edit it
2. "Select" -> Putting the vim cursor on top of a block and making it orange
3. "Highlight" -> Natively selecting block(s) and making them blue

This event is only intended to capture "Focus", and the block does transform into a `textarea` in this case.

I can imagine an `onSelectBlock` event being useful, and potentially enabling lots of features on top of Vim Mode. Is this what you were thinking?

**Comment by tntmarket** [comment]
> Another general comment here - I feel like for commands where we start in `normal` mode we should (most of the time) end up in the normal mode at the end again (even if executing command requires us to go to insert mode temporarily)

I'll investigate making "return to normal mode if we're still in a different mode" an overridable default behavior.

**Comment by tntmarket** [comment]
I extracted functions for the commands. I think we're working towards these layers of logic:

1. **Browser API** - DOM, Keyboard, Mouse, MutationObserver
2. **Roam APIs** - Panel, Highlight, Block, (analogous to the DOM, but with `Roam*` objects/events instead of plain `Element`)
3. **Roam Vim APIs** - "User Space" actions and workflows

I see the functions in `*-commands` as mostly layer 3, with some layer 2 things. I don't know of good abstractions for layer 3 yet, but I think they will emerge as we add/desire more commands.

File: src/ts/core/features/vim-mode/index.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import {ClipboardCommands} from 'src/core/features/vim-mode/commands/clipboard-commands'
+import {PanelCommands} from 'src/core/features/vim-mode/commands/panel-commands'
+import {VisualCommands} from 'src/core/features/vim-mode/commands/visual-commands'
+import {BlockManipulationCommands} from 'src/core/features/vim-mode/commands/block-manipulation-commands'
+import {RoamBlock} from 'src/core/features/vim-mode/roam/roam-block'
+
+export const config: Feature = {
+    id: 'block_navigation_mode',
+    name: 'Vim-like Block Navigation (Requires Refresh)',
+    enabledByDefault: false,
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/index.ts:17]: as this is disabled by default (which probably makes sense at least for now) - I think it'd make sense to add the "warning" message (like in Live Preview) with more details

Comment (Stvad) [reply src/ts/core/features/vim-mode/index.ts:17]: it's not about reliability - I think it's a bit of a switch from the default workflow, so probably should be a conscious choice about enabling this. (also is refresh not required anymore?)

**Comment by Stvad** [comment]
> I extracted functions for the commands. I think we're working towards these layers of logic:
> 
>     1. **Browser API** - DOM, Keyboard, Mouse, MutationObserver
> 
>     2. **Roam APIs** - Panel, Highlight, Block, (analogous to the DOM, but with `Roam*` objects/events instead of plain `Element`)
> 
>     3. **Roam Vim APIs** - "User Space" actions and workflows
> 
> 
> I see the functions in `*-commands` as mostly layer 3, with some layer 2 things. I don't know of good abstractions for layer 3 yet, but I think they will emerge as we add/desire more commands.

This makes sense to me

**Comment by Stvad** [comment]
> > Another general comment here - I feel like for commands where we start in `normal` mode we should (most of the time) end up in the normal mode at the end again (even if executing command requires us to go to insert mode temporarily)
> 
> I'll investigate making "return to normal mode if we're still in a different mode" an overridable default behavior.

Thanks! does not have to be part of this review though

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 9 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-event.ts]: yep. as you mentioned `onSelectBlock` can be useful, but can probably be outside of scope of this PR as we don't use it

File: src/ts/core/features/vim-mode/vim.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+export const returnToNormalMode = async () => {
+    blurEverything()
+    await delay(0)
+    // Clear the native highlight you normally get after blurring a block
+    blurEverything()
+}
+
+type CommandMapper = (key: string, label: string, onPress: (mode: Mode) => void) => Shortcut
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim.ts:35]: I wonder about the interface of passing mode to the handler vs having user define different handlers for different modes. I'd lean to the latter

File: src/ts/core/features/vim-mode/commands/navigation-commands.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import {jumpBlocksInFocusedPanel, Mode, nmap, nvmap} from 'src/core/features/vim-mode/vim'
+import {RoamPanel} from 'src/core/features/vim-mode/roam/roam-panel'
+
+const selectBlockUp = (mode: Mode) => jumpBlocksInFocusedPanel(mode, -1)
+
+const selectBlockDown = (mode: Mode) => jumpBlocksInFocusedPanel(mode, 1)
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/commands/navigation-commands.ts]: it's ok to have cases like these inline :)

File: src/ts/core/features/vim-mode/vim.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    },
+})
+
+export const map: CommandMapper = _map([Mode.NORMAL, Mode.VISUAL, Mode.INSERT])
+export const nmap: CommandMapper = _map([Mode.NORMAL])
+export const nimap: CommandMapper = _map([Mode.NORMAL, Mode.INSERT])
+export const nvmap: CommandMapper = _map([Mode.NORMAL, Mode.VISUAL])
+
+export const jumpBlocksInFocusedPanel = async (mode: Mode, blocksToJump: number) => {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim.ts]: should this be a method on the panel?

File: src/ts/core/features/vim-mode/roam/roam-panel.ts [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        const block = this.selectedBlock().element
+        this.selectBlock(this.relativeBlockId(block.id, blocksToJump))
+    }
+
+    scrollUntilBlockIsVisible(block: BlockElement) {
+        this.scroll(blockScrollOverflow(block))
+    }
+
+    get element(): PanelElement {
+        if (this.panelId === 'SIDE') {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-panel.ts]: this seems like a case for some polymorphism? (have separate classes for main and side panel)

File: src/ts/core/features/vim-mode/roam/roam-panel.ts [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+        return blockId
+    }
+
+    selectedBlock(): RoamBlock {
+        return RoamBlock.get(this.selectedBlockId())
+    }
+
+    selectBlock(blockId: string) {
+        state.panels.set(this.panelId, blockId)
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-panel.ts]: why is this state external to the Panel object?

File: src/ts/core/features/vim-mode/roam/roam-panel.ts [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+        const blocks = this.blocks()
+        const blockIndex = blocks.findIndex(({id}) => id === blockId)
+        return relativeItem(blocks, blockIndex, blocksToJump).id
+    }
+
+    selectedBlockId(): BlockId {
+        const blockId = state.panels.get(this.panelId)
+
+        if (!blockId || !document.getElementById(blockId)) {
+            // Fallback to selecting the first block,
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-panel.ts:59]: is it possible to fallback to something like "closest" (above/below) instead?

File: src/ts/core/features/vim-mode/roam/roam-panel.ts (hunk 16 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-panel.ts]: or we may even get away with just having selector as constructor parameter here.

File: src/ts/core/features/vim-mode/roam/roam-panel.ts (hunk 18 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-panel.ts:59]: maybe by storing an approximate absolute order of last selected block?

File: src/ts/core/features/vim-mode/commands/visual-commands.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
+import {Mode, nmap, nvmap} from 'src/core/features/vim-mode/vim'
+import {Roam} from 'src/core/roam/roam'
+import {Keyboard} from 'src/core/common/keyboard'
+import {RoamBlock} from 'src/core/features/vim-mode/roam/roam-block'
+
+const highlightSelectedBlock = () => Roam.selectBlock(RoamBlock.selected().element)
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/commands/visual-commands.ts]: naming gets confusing here. maybe rename `Roam.selectBlock` to `Roam.highlight` to stay consistent?

File: src/ts/core/features/vim-mode/commands/insert-commands.ts [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+import {Keyboard} from 'src/core/common/keyboard'
+import {RoamBlock} from 'src/core/features/vim-mode/roam/roam-block'
+
+export const insertBlockAfter = async () => {
+    await Roam.activateBlock(RoamBlock.selected().element)
+    await Roam.createBlockBelow()
+}
+
+const editBlock = async () => {
+    await Mouse.leftClick(RoamBlock.selected().element)
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/commands/insert-commands.ts]: I thought the `Roam` had a function for this

**Review by Stvad — COMMENTED** [review]
another partial review, but almost there

File: src/ts/core/features/vim-mode/vim-init.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export const initializeBlockNavigationMode = async () => {
+    await waitForSelectorToExist(Selectors.mainContent)
+    // Wait for roam to finish setting classes on the block, to
+    // avoid classes getting clobbered
+    await delay(300)
+
+    // Select block when clicked
+    RoamEvent.onEditBlock(blockElement => {
+        RoamPanel.select(blockElement.closest(Selectors.mainContent) ? 'MAIN' : 'SIDE')
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim-init.ts]: should probably be static method on panel like `select/getPanelForBlock`

File: src/ts/core/features/vim-mode/vim-view.ts [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    block.classList.add(SELECTED_BLOCK_CSS_CLASS)
+
+    viewMoreDailyLogIfPossible()
+
+    return null
+}
+
+const clearHighlights = () => {
+    const priorSelections = document.querySelectorAll(`.${SELECTED_BLOCK_CSS_CLASS}`)
+    if (priorSelections.length > 0) {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim-view.ts]: this check seems redundant

File: src/ts/core/features/vim-mode/vim-view.ts [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        priorSelections.forEach(selection => selection.classList.remove(SELECTED_BLOCK_CSS_CLASS))
+    }
+}
+
+const viewMoreDailyLogIfPossible = () => {
+    const viewMore = document.querySelector(Selectors.viewMore)
+    if (viewMore && isElementVisible(viewMore)) {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim-view.ts]: wonder if `isElementVisible` should accept nullable things and return false if input is null/undefined

File: src/ts/core/features/vim-mode/vim-view.ts [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * Runs side effects such as highlighting the selected block,
+ * in order to update Roam to reflect Vim Mode's internal state
+ */
+
+const BLUR_PIXEL = 'roam-toolkit-block-mode--unfocus-pixel'
+const SELECTED_BLOCK_CSS_CLASS = 'roam-toolkit-block-mode--highlight'
+injectStyle(
+    `
+    #${BLUR_PIXEL} {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/vim-view.ts]: it has to be visible for us to be able to click it?

File: src/ts/core/features/vim-mode/vim.ts (hunk 15 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/vim.ts]: I was avoiding `Panel` calling vim things like `Mode`, so `Panel` would be more reusable. Maybe extract a class like `RoamPanelVim`?

Comment (Stvad) [reply src/ts/core/features/vim-mode/vim.ts]: Hmm, that's fair. sure, that class sounds reasonable

File: src/ts/core/features/vim-mode/roam/roam-panel.ts (hunk 16 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-panel.ts]: I personally prefer `RoamPanel`'s interface gloss over the side/main distinction (using the abstract notion of `PanelId`) because:

* The interface is easier to use if it hides the difference between main/side, and doesn't require data that's technically derivable from `PanelId`
* Main/side panels seem pretty similar from a UX perspective. I'm guessing other special cases will be rare. Differing selectors feels like an implementation detail addressing Roam's DOM structure, rather than an essential distinction between main/side to model.
* I think it's theoretically possible to turn panels into truly generic and multiplexed vim windows, and navigate between them spatially using [elementFromPoint](https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/elementFromPoint), which may conflict with having variations of panels.

File: src/ts/core/features/vim-mode/roam/roam-panel.ts (hunk 17 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-panel.ts]: I guess I'm used to using top level variables and functions, more than classes :p
I'll move this to be a static private variable.

File: src/ts/core/features/vim-mode/roam/roam-panel.ts (hunk 18 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-panel.ts:59]: Filed https://github.com/roam-unofficial/roam-toolkit/issues/100

File: src/ts/core/features/vim-mode/vim-view.ts (hunk 24 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/vim-view.ts]: Oh apparently not, I'll remove this css 😛

File: src/ts/core/features/vim-mode/vim.ts (hunk 13 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/vim.ts:35]: I think being separate handlers for modes is better when the modes don't share code:

```typescript
nmap('k', 'Select Block Up', () => RoamPanel.selected().selectRelativeBlock(-1)),
vmap('k', 'Select Block Up', () => RoamHighlight.selectRelative(-1)),
nmap('j', 'Select Block Down', () => RoamPanel.selected().selectRelativeBlock(1)),
vmap('j', 'Select Block Up', () => RoamHighlight.selectRelative(1)),
```

There's a few handlers such as `copySelectedBlock` where the modes share code.

I think having both would be the nicest interface.

The main reason I didn't do this was cause I'd have to merge handlers that use the same key into one handler. Otherwise, `react-hotkeys` only binds the first one.

File: src/ts/core/features/vim-mode/index.ts (hunk 12 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/index.ts:17]: Got it, I'll say _"Warning: Intrusive, may interfere with your regular workflow"_ then.

I was testing earlier, and the highlight seems to appear after I check the box and waiting a little bit so 🤷

File: src/ts/core/features/vim-mode/vim.ts (hunk 13 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/vim.ts:35]: I filed https://github.com/roam-unofficial/roam-toolkit/issues/101 to remember refactorings that'll make it easier to develop on top of Vim mode

**Comment by tntmarket** [comment]
I submitted some small fixes, I'm still thinking about how to address https://github.com/tntmarket/roam-toolkit/pull/3#discussion_r449382695 and https://github.com/tntmarket/roam-toolkit/pull/3#discussion_r449380786

[Comparison of different data models for panel state](https://gist.github.com/tntmarket/eef674b604d5e081e46def812a5efeef)

Quote from Slack:
---
Hmm ok, so I just remembered that I use top level Panel.focusedBlock state in https://github.com/tntmarket/roam-toolkit/pull/5/files, because the code:

* Observes when sidebar pages are added/removed
* Whenever the sidebar changes, refresh the panel state to match all the currently displayed pages
* The top level state remembering PanelId => BlockId is reused

If I store the focused block on panel instances, then I need to either re-hydrate the `focusedBlock` for each panel instance, or identify the specific panel being removed/inserted when synchronizing panel state - [example code](https://gist.github.com/tntmarket/dcf014f83cacccca52516bc86e43bff4)

The basic PR for core Vim functionality doesn’t need this, since it assumes only 1 main/side panel, but I storing panels in a list and moving forward/backwards in the list enables future features like jumping between sidebar pages.

File: src/ts/core/features/vim-mode/vim.ts (hunk 13 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/vim.ts:35]: shared code can be extracted into functions/etc.  🤷‍♂️

>hmm, just looked at the code, I’m not sure I understand your concern - there you still have the 1-1 panel to block mapping 

ah, that's unfortunate, but should be fairly straightforward given that we do prepossessing on them already. but let's keep that outside of this PR

**Comment by tntmarket** [comment]
I made a few changes to how Panels are persisted:

* Have `RoamPanel` own which block is selected
* Inject element into `RoamPanel`, rather than having `RoamPanel` look for it's own element
* Use the `PanelElement` as a unique identifier for Panels
* Navigate left/right across a list of `PanelId[]`, rather than flipping between hardcoded main/side panels
* Persist a map of `PanelId` => `RoamPanel`, rather than a map from `PanelId` => `Focused Block`

File: src/ts/core/features/vim-mode/vim.ts (hunk 15 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/vim.ts]: 👀

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 6 above)
Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-event.ts:33]: 👀

File: src/ts/core/features/vim-mode/roam/roam-panel.ts [hunk 25]
```diff
@@ -95,20 +92,46 @@ export class RoamPanel {
...
+    static fromBlock(blockElement: BlockElement): RoamPanel {
+        const panelElement = assumeExists(blockElement.closest(PANEL_SELECTOR)) as PanelElement
+        return assumeExists(state.panels.get(panelElement))
     }
 
-    static selectPanelContainingBlock(blockElement: BlockElement) {
-        if (blockElement.closest(Selectors.mainContent)) {
-            return RoamPanel.select('MAIN')
-        }
+    static at(panelIndex: PanelIndex): RoamPanel {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-panel.ts]: not really sure if we should be exposing this

File: src/ts/core/features/vim-mode/roam/roam-panel.ts [hunk 26]
```diff
@@ -95,20 +92,46 @@ export class RoamPanel {
...
+    }
+
+    static updateSidePanels() {
+        tagPanels()
+        state.panelOrder = Array.from(document.querySelectorAll(PANEL_SELECTOR)) as PanelElement[]
+        state.panelOrder.forEach(RoamPanel.persistPanel)
+    }
+
+    private static persistPanel(element: PanelElement) {
+        if (!state.panels.has(element)) {
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/roam/roam-panel.ts]: I think this one is a potential memory leak, as we'd be constantly adding things to the map without ever removing them. Probably not a big deal, but potentially better to re-create map?

**Review by Stvad — APPROVED** [review]
Looks good to me in general! but take a look at the potential memory leak comment

File: src/ts/core/features/vim-mode/commands/history-commands.ts [hunk 27]
```diff
@@ -0,0 +1,15 @@
+import {nmap, returnToNormalMode} from 'src/core/features/vim-mode/vim'
+import {Keyboard} from 'src/core/common/keyboard'
+import {KEY_TO_CODE} from 'src/core/common/keycodes'
+
+const undo = async () => {
+    await Keyboard.simulateKey(KEY_TO_CODE['z'], 0, {key: 'z', metaKey: true})
```
Comment (Stvad) [inline src/ts/core/features/vim-mode/commands/history-commands.ts:6]: why do we need both key and key code?

File: src/ts/core/features/vim-mode/roam/roam-event.ts (hunk 6 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/roam/roam-event.ts:33]: By defined type, do you mean the existing `DisconnectFn` definition?

Comment (Stvad) [reply src/ts/core/features/vim-mode/roam/roam-event.ts:33]: uh, never-mind this one, for some reason I read it as declaration and not an assignment

File: src/ts/core/features/vim-mode/commands/history-commands.ts (hunk 27 above)
Comment (tntmarket) [reply src/ts/core/features/vim-mode/commands/history-commands.ts:6]: I'm not sure TBH, this happens to be the combination of parameters that worked. If I remove either, the simulated undo/redo doesn't work anymore

**Comment by tntmarket** [comment]
Regarding `onRightPanelChange`, did you mean to use the already defined type for `DisconnectFn`? (I switched to using the existing typedef already)

**Comment by Stvad** [comment]
👍
